### PR TITLE
Value Server

### PR DESCRIPTION
**Pull request by gpauloski**
**Introduces:**
- Value Server class for interfacing with Redis (with caching and limited object version guarantees)
- Transparent Python Object Proxy class for abstracting interfacing with objects in the value server
- Demo synthetic data app for benchmarking

**Usage:**
Colmena will automatically scan task inputs/outputs and pass objects via the value server based on a user defined threshold size.
```
client_queues, server_queues = make_queue_pairs(
    args.redis_host,
    args.redis_port,
    topics=['generate'],
    serialization_method='pickle',
    value_server_threshold=100000    # pass objects larger than 100KB via the value server
) 
```
By default, the Redis server used for the queues will also be used for the Value Server. Alternatively, a `value_server_hostname` and `value_server_port` can be passed to `make_queue_pairs()`.

Remote workers connect to the Value Server via environment variables. The environment variables can be set via the `worker_init` in Parsl's `Provider` class.
```
from colmena import value_server
...
LocalProvider(
   ...
    worker_init='export {}={}\nexport {}={}\n'.format(
        value_server.VALUE_SERVER_HOST_ENV_VAR, args.redis_host,
        value_server.VALUE_SERVER_PORT_ENV_VAR, args.redis_port)
)
...
```

### CODE REVIEW

File: colmena/value_server/server.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        """
+        self.redis_client = redis.StrictRedis(
+            host=hostname, port=port, decode_responses=True)
+        self.cache = LRUCache(LRU_CACHE_SIZE)
+
+    def is_cached(self, key: str, strict: bool = False) -> bool:
+        """Check if key is cached locally
+
+        Args:
+            key (str)
```
Comment (WardLT) [inline colmena/value_server/server.py:77]: Could you explain what key is?

File: colmena/value_server/proxy.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            most recent object associated with this key
+
+    Returns:
+        Object or iterable with same format as `objs` with any objects of
+        `sys.getsizeof` greater than `threshold` replace by proxies.
+    """
+    def _to_proxy(obj: Any) -> Any:
+        if isinstance(obj, ObjectProxy) or threshold is None:
+            return obj
+        if sys.getsizeof(obj) > threshold:
```
Comment (WardLT) [inline colmena/value_server/proxy.py:160]: Could we put some debug-level logging in here so we can track which objects are getting proxied?

File: colmena/value_server/server.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            obj (object)
+            serialization_method (SerializationMethod): serialization method
+                to use for serializing object before putting in value server
+        """
+        value = SerializationMethod.serialize(serialization_method, obj)
+        self.redis_client.set(key, value)
+        self.redis_client.set(key + '_timestamp', time.time())
+
+
+def init_value_server(hostname: Optional[str] = None,
```
Comment (WardLT) [inline colmena/value_server/server.py:139]: Environment variables makes me a bit nervous. Do you think we should inject the connection information for the server into the result objects?

[Note: Not a blocker for merging]

File: colmena/value_server/__init__.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+from colmena.value_server.proxy import to_proxy
+from colmena.value_server.proxy import to_proxy_threshold
+from colmena.value_server.proxy import ObjectProxy
+
+from colmena.value_server.server import init_value_server
+from colmena.value_server.server import LRUCache
+from colmena.value_server.server import VALUE_SERVER_HOST_ENV_VAR
+from colmena.value_server.server import VALUE_SERVER_PORT_ENV_VAR
+
+global server
```
Comment (WardLT) [inline colmena/value_server/__init__.py:13]: Another thing to eventually change down the road: Allowing more multiple servers

File: colmena/value_server/proxy.py (hunk 2 above)
Comment (WardLT) [inline colmena/value_server/proxy.py:160]: Do you know if sizeof works well with tensorflow models or other weird objects? If we run into problems, can we run this "to_proxy" after serializing an object?

**Review by WardLT — APPROVED** [review]
Looks great! I think it is ready to merge.

I have listed a few comments around ideas I think we could improve later, especially around configuration. We at least need something to automate setting the configuration for the workers. Maybe a "ValueServerConfig" object that we give to the MethodServer that will either pass those details along with the task object, or use them to modify the Parsl configuration. Not yet sure whats best...

My other thought is we need to document somewhere (perhaps in the readthedocs) how all of this works and should be used. I am starting to write up how "multi-site" and the dynamic resource allocation methods work and are used, and will point you to those docs once I'm done.

**Comment by WardLT** [comment]
I decided to go-ahead and merge this as I'm making an overhaul of other aspects of the code. 

Let's open issues to capture some of these thoughts.

**Comment by gpauloski** [comment]
Sounds good. I will go ahead and open the relevant issues and get started on them once I'm done with classes.

**Pull request merged**